corehq/pillows/user.py

```diff
@@ -41,6 +41,24 @@ class UserPillow(AliasedElasticPillow):
     def get_unique_id(self):
         return USER_INDEX
 
+    def change_transform(self, doc_dict):
+        super(UserPillow, self).change_transform(doc_dict)
+        doc_dict = self._cast_user_data_to_string(doc_dict)
+        return doc_dict
+
+    def _cast_user_data_to_string(self, doc_dict):
+        """
+        Make all user_data strings
+        ES doesn't allow dynamic dicts with the same key name to have different types, so coerce everything
+        """
+        user_data = doc_dict.get('user_data', {})
+        if user_data and type(user_data) is dict:
+            try:
+                doc_dict['user_data'] = {k: unicode(v) for k, v in user_data.iteritems()}
+            except UnicodeDecodeError:
+                pass            # If we can't decode it, let elastic deal with it
```

Comment:
Would it make sense to convert all values that you can, and only ignore byte strings that can't be converted. Example:

`{u"number": 1, u"bytes": b"\xe4\x94"}`

should be converted to

`{u"number": u"1", u"bytes": b"\xe4\x94"}`

The code here will skip conversion entirely if a single value cannot be converted.